### ISSUE TITLE
API Management - Cannot add a revision to an existing API

### ISSUE THREAD

**Issue by Berbe** [hashicorp/terraform-provider-azurerm#12720]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
```console
Terraform v1.0.3
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.69.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
```

### Affected Resource(s)
* `azurerm_api_management_api`

### Terraform Configuration Files
Based on an existing revision:
```hcl
resource "azurerm_api_management_api" "example" {
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  display_name        = "example"
  path                = "example"
  revision            = "1"
  protocols           = [
    "https",
  ]
  service_url         = "https://example.org"

  import {
    content_format = "openapi+json"
    content_value  = file("${path.module}/example.json")
  }
}
```
, create a new one:
```hcl
resource "azurerm_api_management_api" "example_2" {
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  display_name        = "example"
  path                = "example"
  revision            = "2"
  protocols           = [
    "https",
  ]
  service_url         = "https://example.org"

  import {
    content_format = "openapi+json"
    content_value  = file("${path.module}/example.json")
  }
}
```

### Expected Behaviour
Revision `2` created alongside `1`

### Actual Behaviour
```console
Error: waiting on creating/updating API Management API "example" (Resource Group "example"): Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=400 -- Original Error: Code="ValidationError" Message="Can't change property Name for non-current revision "
```

### Steps to Reproduce
1. `terraform apply`

### Additional information
If you try to create both at once, ie with:
```hcl
resource "azurerm_api_management_api" "example" {
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  display_name        = "example"
  path                = "example"
  revision            = "1"
  protocols           = [
    "https",
  ]
  service_url         = "https://example.org"

  import {
    content_format = "openapi+json"
    content_value  = file("${path.module}/example.json")
  }
}

resource "azurerm_api_management_api" "example_2" {
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  display_name        = "example"
  path                = "example"
  revision            = "2"
  protocols           = [
    "https",
  ]
  service_url         = "https://example.org"

  import {
    content_format = "openapi+json"
    content_value  = file("${path.module}/example.json")
  }
}
}
```
, it works!
Revisions `1` & `2` are created.

**Comment by nrgraghunath** [comment]
@Berbe  I think its working as expected,  as per code it'll forced as a new resource to be created. if you change a version rather than updating the existing api resource.

**Comment by yupwei68** [comment]
Hi @Berbe , thanks for opening this issue.
I think you could create a new revision from an existing revision like the example here https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/apimanagement/api_management_api_resource_test.go#L664.
```
resource "azurerm_api_management_api" "revision" {
  name                 = "acctestRevision-%d"
  resource_group_name  = azurerm_resource_group.test.name
  api_management_name  = azurerm_api_management.test.name
  revision             = "18"
  source_api_id        = "${azurerm_api_management_api.test.id};rev=3"
  revision_description = "Creating a Revision of an existing API"
}
```

**Comment by nrgraghunath** [comment]
@yupwei68 Exactly, @Berbe please try as per the example defined here. It works. Name validation restricts to provision the resources so please use the example as the link https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/apimanagement/api_management_api_resource_test.go#L664

**Comment by Berbe** [comment]
If I try the following:
```hcl
resource "azurerm_api_management_api" "example-2" {
  name = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  revision = "2"
  service_url = "http://example.org"
  source_api_id = "${azurerm_api_management_api.example.id};rev=1"

  import {
    content_format = "openapi+json"
    content_value = file("${path.module}/example.json")
  }
}
```
I end up with:
```console
Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=400 -- Original Error: Code="ValidationError" Message="One or more fields contain incorrect values:" Details=[{"code":"ValidationError","message":"Cannot create API 'example' with the same Path '' as API '***'  unless it's a part of the same version set","target":"path"}
```
`source_api_id` hence does not replace the `path` property.

If I try:
```hcl
resource "azurerm_api_management_api" "example-2" {
  name = "example_2"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  revision = "2"
  service_url = "http://example.org"
  source_api_id = "${azurerm_api_management_api.example.id};rev=1"
  path = "example"

  import {
    content_format = "openapi+json"
    content_value = file("${path.module}/example.json")
  }
}
```
I end up with:
```console
Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=400 -- Original Error: Code="ValidationError" Message="One or more fields contain incorrect values:" Details=[{"code":"ValidationError","message":"Cannot create API 'example_2' with the same Path 'example' as API 'example'  unless it's a part of the same version set","target":"path"}]
```

Of course, trying:
```hcl
resource "azurerm_api_management_api" "example-2" {
  name = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  revision = "2"
  service_url = "http://example.org"
  source_api_id = "${azurerm_api_management_api.example.id};rev=1"
  path = "example"

  import {
    content_format = "openapi+json"
    content_value = file("${path.module}/example.json")
  }
}
```
gives me the exacte same error from my original post:
```console
Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=400 -- Original Error: Code="ValidationError" Message="Can't change property Name for non-current revision "
```

_`source_api_id` does not seem to be of any help here as I eventually fall back on the original error, and it is kind of a relief anyway, as it's an optional directive. As my tests show, this directive does not even seem to be self-sufficient to fully describe an API based on another one.
Maybe a bug report shall be filled for this defect, but this is beyond the scope of this particular issue. @yupwei68 If you're interested in it, I'll let you deal with it._

I am so confused...

Trying to sum up all that has been tried and advised goes down to this: defining revisions all-at-once or separately produce different results.
This feels wrong, as Terraform configuration is supposed to be deterministic, not contextually based: it should use its own state to decide what needs to be done to reach the aforementioned determined state.
Said otherwise, the same configuration applied to different contexts might involve different actions, but that's the deal with deterministic tools: it should handle variations themselves.

Going back to the original issue description, how come the same configuration applied to different contexts does not yield the same results?

**Comment by Berbe** [comment]
OK I think I got a lead.

The problematic part was the `import` block.
Without it, I am able to create a revision:
```hcl
resource "azurerm_api_management_api" "example_2" {
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
  api_management_name = azurerm_api_management.example.name
  revision            = 2
  path                = "example"
  service_url         = "http://example.org"
  source_api_id       = "${azurerm_api_management_api.example.id};rev=1"
}
```

The problem here is that, depending on whether you are creating the first revision of an API or not, the configuration you shall provide needs to be different.
This hurts automated Terraform configuration generation.

**Comment by axk511** [comment]
I have exact issue as we are trying to generate the terraform code automatically - similar to @Berbe mentioned above. 
Was wondering if this can be fixed?

**Comment by axk511** [comment]
Issue here is revisions are supposed to help us non-breaking changes to API's - like adding an operation to existing API.
since the current version of terraform errors out when using below section, it defeats the purpose of revisions.
import {
    content_format = "openapi+json"
    content_value  = file("${path.module}/example.json")
  }

Here is the link to revisions from microsoft docs
https://docs.microsoft.com/en-us/azure/api-management/api-management-get-started-revise-api?tabs=azure-portal

**Comment by Firenza** [comment]
We are also running into this same issue.  Not being able to use revisions a big pain point for us 😞

**Comment by aaronmottmac** [comment]
My team is also running into this same issue

**Comment by ZeChArtiahSaher** [comment]
Yeah, please provide some actual examples of your multi-state pattern which should be conforming to simple terraform eventual state model

**Comment by Firenza** [comment]
My team is also running into this issue.

**Comment by mnemcik-visma** [comment]
Hi, could we get any status update to this issue and indication whether it's planned for fixing/implementation? The issue essentially renders Revisions functionality unusable from Terraform

**Comment by taksan** [comment]
This issue still affects revision up to this date

**Comment by Apostolos-Daniel** [comment]
What's the latest on this issue? I've reproduced the fact that revisions cannot be created (and updated to current). Are there plans to implement this?

See [example code](https://github.com/Apostolos-Daniel/azure-apim-terraform/blob/main/README.md#revisions)

**Comment by rcskosir** [comment]
Updating PR linking for this issue - PR #22380 has been superseded by #23031.
